### PR TITLE
Serve easily dynamic files with `DataFromReader` context method

### PR DESCRIPTION
**Pull request by jclebreton**
This pull request adds a new method called `DataFromReader` to provides an easy way to serve dynamic files.

### CODE REVIEW

**Comment by appleboy** [comment]
@thinkerou Could you also help to review this PR?

**Comment by thinkerou** [comment]
@appleboy OK~

File: README.md [hunk 1]
```diff
@@ -834,6 +834,32 @@ func main() {
 }
 ```
 
+### Serving data from reader
```
Comment (thinkerou) [inline README.md]: Please add content link in README, thanks!

File: render/reader.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	ContentType   string
+	ContentLength int64
+	Reader        io.Reader
+	Headers       map[string]string
+}
+
+// Render (Reader) writes data with custom ContentType and headers.
+func (r Reader) Render(w http.ResponseWriter) (err error) {
+	r.WriteContentType(w)
+	r.Headers["Content-Length"] = fmt.Sprintf("%d", r.ContentLength)
```
Comment (thinkerou) [inline render/reader.go]: Why not use `strconv.Itoa(r.ContentLength)`? Thanks!

File: README.md (hunk 1 above)
Comment (jclebreton) [reply README.md]: Hello @thinkerou I'm not sure to understand the target of the asked link. This page ?

File: render/reader.go (hunk 2 above)
Comment (jclebreton) [reply render/reader.go]: You're right. It's better to use `strconv` library but to convert a int64 variable to a string  `strconv.FormatInt(r.ContentLength, 10)` works better ;)

File: README.md (hunk 1 above)
Comment (thinkerou) [reply README.md]: You should add `[Serving data from reader](serving-data-from-reader)` after the line:
image

File: render/reader.go (hunk 2 above)
Comment (thinkerou) [reply render/reader.go]: Good!

File: render/render_test.go [hunk 3]
```diff
@@ -11,8 +11,11 @@ import (
 	"html/template"
 	"net/http"
 	"net/http/httptest"
+	"strconv"
 	"testing"
 
+	"strings"
```
Comment (appleboy) [inline render/render_test.go]: remove the extra empty line between internal package.

Comment (jclebreton) [reply render/render_test.go]: Done ;-)

**Comment by thinkerou** [comment]
LGTM

File: render/reader.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+func (r Reader) WriteContentType(w http.ResponseWriter) {
+	writeContentType(w, []string{r.ContentType})
+}
+
+func (r Reader) writeHeaders(w http.ResponseWriter, headers map[string]string) {
+	header := w.Header()
+	for k, v := range headers {
+		if val := header[k]; len(val) == 0 {
```
Comment (thinkerou) [inline render/reader.go]: Actually, I prefer to using `val == ""` for judging empty string.

Comment (appleboy) [reply render/reader.go]: @jclebreton Please update and I will merge this.

**Comment by jclebreton** [comment]
Thanks for review and merge ;-)

**Comment by LoveSnowBear** [comment]
@jclebreton May i ask a question, if i call context.dataFromReader function, when do i close reader or it doesn't need to close?

**Pull request merged**